pym/_emerge/depgraph.py

```diff
@@ -3584,6 +3584,15 @@ def _minimize_children(self, parent, priority, root_config, atoms):
 						if atom_set.findAtomForPackage(pkg2, modified_use=self._pkg_use_enabled(pkg2)):
 							atom_pkg_graph.add(pkg2, atom)
 
+			# In order for the following eliminate_pkg loop to produce
+			# deterministic results, the order of the pkgs list must
+			# not be random (bug 631894). Prefer to eliminate installed
+			# packages first, in case rebuilds are needed, and also sort
+			# in ascending order so that older versions are eliminated
+			# first.
+			pkgs = (sorted(pkg for pkg in pkgs if pkg.installed) +
```

Comment:
How big is the list going to be? Maybe it makes sense to bisect.insort here and not touch every element twice.
```
installed = []
not_installed = []
for pkg in pkgs:
  if pkg.installed:
    bisect.insort(installed, pkg)
  else:
    bisect.insort(not_installed,pkg)

pkgs = installed + not_installed
```

Reply:
Typically we're only dealing with 2 packages, so there's not much cost here to optimize away. Also, insort would be less efficient than something like this:
```python
installed = []
not_installed = []
for pkg in sorted(pkgs):
   if pkg.installed:
     installed.append(pkg)
   else:
     not_installed.append(pkg)
```
And that's going to have a negligible performance difference compared to my 2-line version that's easier to read.

Reply:
okay, if it's just a small list then nevermind my comment 👍